cpu/esp32/periph/pm.c

```diff
@@ -41,65 +44,127 @@ void pm_set_lowest(void)
     system_wdt_feed();
     #endif
 
-    DEBUG ("%s exit from sleep @%u\n", __func__, system_get_time());
+    DEBUG ("%s exit from normal sleep @%u\n", __func__, system_get_time());
 }
 
 void IRAM_ATTR pm_off(void)
+{
+    /* enter hibernate mode without any enabled wake-up sources */
+    esp_deep_sleep_start();
+}
+
+extern void esp_restart_noos(void) __attribute__ ((noreturn));
+
+void pm_reboot(void)
 {
     DEBUG ("%s\n", __func__);
 
-    /* suspend UARTs */
+    /* suspend and flush UARTs */
     for (int i = 0; i < 3; ++i) {
         REG_SET_BIT(UART_FLOW_CONF_REG(i), UART_FORCE_XOFF);
         uart_tx_wait_idle(i);
     }
 
-    /* set all power down flags */
-    uint32_t pd_flags = RTC_SLEEP_PD_DIG |
-                        RTC_SLEEP_PD_RTC_PERIPH |
-                        RTC_SLEEP_PD_RTC_SLOW_MEM |
-                        RTC_SLEEP_PD_RTC_FAST_MEM |
-                        RTC_SLEEP_PD_RTC_MEM_FOLLOW_CPU |
-                        RTC_SLEEP_PD_VDDSDIO;
+    software_reset();
+}
+
+#ifndef MODULE_PM_LAYERED
 
-    rtc_sleep_config_t config = RTC_SLEEP_CONFIG_DEFAULT(pd_flags);
-    config.wifi_pd_en = 1;
-    config.rom_mem_pd_en = 1;
-    config.lslp_meminf_pd = 1;
+void pm_set_lowest(void)
+{
+    pm_set_lowest_normal();
+}
 
-    /* Save current frequency and switch to XTAL */
-    rtc_cpu_freq_t cpu_freq = rtc_clk_cpu_freq_get();
-    rtc_clk_cpu_freq_set(RTC_CPU_FREQ_XTAL);
+#else /* MODULE_PM_LAYERED */
 
-    /* set deep sleep duration to forever */
-    rtc_sleep_set_wakeup_time(rtc_time_get() + ~0x0UL);
+void pm_set(unsigned mode)
+{
+    if (mode == ESP_PM_MODEM_SLEEP) {
+        pm_set_lowest_normal();
+        return;
+    }
 
-    /* configure deep sleep */
-    rtc_sleep_init(config);
-    rtc_sleep_start(RTC_TIMER_TRIG_EN, 0);
+    DEBUG ("%s enter to power mode %d @%u\n", __func__, mode, system_get_time());
 
-    /* Restore CPU frequency */
-    rtc_clk_cpu_freq_set(cpu_freq);
+    /* Labels for RTC slow memory that are defined in the linker script */
+    extern int _rtc_bss_rtc_start;
+    extern int _rtc_bss_rtc_end;
 
-    /* resume UARTs */
-    for (int i = 0; i < 3; ++i) {
-        REG_CLR_BIT(UART_FLOW_CONF_REG(i), UART_FORCE_XOFF);
-        REG_SET_BIT(UART_FLOW_CONF_REG(i), UART_FORCE_XON);
-        REG_CLR_BIT(UART_FLOW_CONF_REG(i), UART_FORCE_XON);
+    /*
+     * Activate the power domain for the slow RTC memory when there are
```

Comment:
s/there are/there is

Reply:
Forget it. Is seems to be very disputable: https://en.wikipedia.org/wiki/Data_(word)

Reply:
I will use another sentence which avoids the problem with countable nouns.